pkg/kubelet/pod_workers.go

```diff
@@ -221,76 +602,412 @@ func (p *podWorkers) UpdatePod(options *UpdatePodOptions) {
-		// since per-pod goroutine won't be able to put it to the already closed
-		// channel when it finishes processing the current work update.
-		delete(p.lastUndeliveredWorkUpdate, uid)
+// calculateEffectiveGracePeriod sets the initial grace period for a newly terminating pod or allows a
+// shorter grace period to be provided, returning the desired value.
+func calculateEffectiveGracePeriod(status *podSyncStatus, pod *v1.Pod, options *KillPodOptions) (int64, bool) {
+	// enforce the restriction that a grace period can only decrease and track whatever our value is,
+	// then ensure a calculated value is passed down to lower levels
+	gracePeriod := status.gracePeriod
+	// this value is bedrock truth - the apiserver owns telling us this value calculated by apiserver
+	if override := pod.DeletionGracePeriodSeconds; override != nil {
+		if gracePeriod == 0 || *override < gracePeriod {
+			gracePeriod = *override
+		}
+	}
+	// we allow other parts of the kubelet (namely eviction) to request this pod be terminated faster
+	if options != nil {
+		if override := options.PodTerminationGracePeriodSecondsOverride; override != nil {
+			if gracePeriod == 0 || *override < gracePeriod {
+				gracePeriod = *override
+			}
+		}
+	}
+	// make a best effort to default this value to the pod's desired intent, in the event
+	// the kubelet provided no requested value (graceful termination?)
+	if gracePeriod == 0 && pod.Spec.TerminationGracePeriodSeconds != nil {
+		gracePeriod = *pod.Spec.TerminationGracePeriodSeconds
+	}
+	// no matter what, we always supply a grace period of 1
+	if gracePeriod < 1 {
+		gracePeriod = 1
+	}
+	return gracePeriod, status.gracePeriod != 0 && status.gracePeriod != gracePeriod
+}
+
+func (p *podWorkers) managePodLoop(podUpdates <-chan podWork) {
+	var lastSyncTime time.Time
+	var lastKillTime time.Time
+	for update := range podUpdates {
+		pod := update.Options.Pod
+
+		klog.V(4).InfoS("DEBUG: Processing pod event", "pod", klog.KObj(pod), "podUID", pod.UID, "updateType", update.WorkType)
+		err := func() error {
+			// The worker is responsible for ensuring the sync method sees the appropriate
+			// status updates on resyncs (the result of the last sync), transitions to
+			// terminating (no wait), or on terminated (whatever the most recent state is).
+			// Only syncing and terminating can generate pod status changes, while terminated
+			// pods ensure the most recent status makes it to the api server.
+			var status *kubecontainer.PodStatus
+			var err error
+			switch {
+			case update.Options.RunningPod != nil:
+				// when we receive a running pod, we don't need status at all
+			case update.WorkType == TerminatingPodWork:
+				if lastKillTime.IsZero() {
+					// if this is the first time we are killing the pod, we don't want to wait on the pod
+					// cache at all as it may be out of date and we want to begin termination as quickly
+					// as possible
+					lastKillTime = time.Now()
+					status = nil
```

Comment:
So I debated that - but the use case for terminating is

a) i need to ensure i have the latest for killPod - i prefer to have that code close to the call (not buried in the pod worker) so that it's clear that you need the latest state for kill pod
b) after we kill the pod we MUST know that all containers are stopped - a crio bug is fatal and should be caught, and we ALSO want to ensure the latency is low between kill (when no containers can run) and when we refresh the cache (so that if containers are purged, we have the status up to date so we don't lose the container status)

So for those reasons I prefer to keep the status refresh close to the kill, even though I think it's cleaner as you are describing.

Reply:
The new commit LGTM.